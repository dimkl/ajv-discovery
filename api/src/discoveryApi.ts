import type Ajv from "ajv";
import type { AnySchema } from "ajv";

type DataType = Record<string, unknown> | null | undefined;
type DataAccessor = () => DataType | Promise<DataType>;
type DataSetter = (data: DataType) => void;

type ValidateSchemaType = {
  schema: AnySchema;
  dataAccessor: DataAccessor;
  dataSetter?: DataSetter;
};

export class DiscoveryApi {
  readonly ajv: Ajv;

  constructor(ajv: Ajv) {
    this.ajv = ajv;
  }

  getSchemas() {
    const schemas = Object.keys(this.ajv.schemas)
      .filter((s) => s.startsWith("/schemas/"))
      .map((s) => this.ajv.getSchema(s)!.schema);

    return { schemas };
  }

  registerSchema(schema: AnySchema, silent = false) {
    this.ajv.addSchema(schema);
    !silent && this.ajv.validateSchema(schema);
  }

  async validateSchema({
    schema,
    dataAccessor,
    dataSetter = async () => {},
  }: ValidateSchemaType) {
    const validate = this.ajv.compile(schema);
    const data = await dataAccessor();
    if (await validate(data)) {
      await dataSetter(data);
    }
  }
}
