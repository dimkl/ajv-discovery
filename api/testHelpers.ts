export function createRequiredError(attrName: string) {
    return {
      instancePath: "",
      keyword: "required",
      message: `must have required property '${attrName}'`,
      params: {
        missingProperty: attrName,
      },
      schemaPath: "#/required",
    };
  }
  