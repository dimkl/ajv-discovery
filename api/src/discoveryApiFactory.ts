import Ajv from "ajv";
import addFormats from "ajv-formats";

import { addKeywords } from "./addKeywords";
import { DiscoveryApi } from "./discoveryApi";

let cache: Record<string, Ajv> = {};
const DEFAULT_KEY = "default";

export class DiscoveryApiFactory {
  static getInstance(key: string = DEFAULT_KEY, ajv?: Ajv) {
    if (!cache[key]) {
      cache[key] = ajv || this.createAjv();
    }

    if (ajv && ajv !== cache[key]) {
      cache[key] = ajv;
    }

    return new DiscoveryApi(cache[key]);
  }

  static clear() {
    cache = {};
  }

  static createAjv() {
    const ajvInstance = new Ajv({
      allErrors: true,
      removeAdditional: true,
    });

    addKeywords(ajvInstance);
    addFormats(ajvInstance);

    return ajvInstance;
  }
}
