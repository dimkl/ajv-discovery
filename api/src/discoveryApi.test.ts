import * as schema from "../fixtures/customer.create.json";
import * as userSchema from "../fixtures/user.json";

import { createRequiredError } from "../testHelpers";

import { DiscoveryApi } from "./discoveryApi";
import { DiscoveryApiFactory } from "./discoveryApiFactory";

describe("DiscoveryApi", () => {
  describe("getSchemas()", () => {
    test("returns all registered schemas {schemas: AnySchema[]}", () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      expect(discoveryApi.getSchemas()).toMatchSnapshot();

      discoveryApi.registerSchema(schema);
      const schema2 = { ...schema, $id: "/schemas/customer.update" };
      discoveryApi.registerSchema(schema2);
      expect(discoveryApi.getSchemas()).toMatchSnapshot();
    });

    test("returns all non /schemas.* ids as {schemas: AnySchema[]}", () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      expect(discoveryApi.getSchemas()).toMatchSnapshot();

      discoveryApi.registerSchema(schema);
      const schema2 = { ...schema, $id: "/schemas/customer.update" };
      discoveryApi.registerSchema(schema2);
      // Should not be in getSchemas
      const schema3 = { ...schema, $id: "/excluded" };
      discoveryApi.registerSchema(schema3);
      expect(discoveryApi.getSchemas()).toMatchSnapshot();
    });
  });

  describe("registerSchema(schema)", () => {
    test("throws error for already registered schema", () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      discoveryApi.registerSchema(schema);
      expect(() => discoveryApi.registerSchema(schema)).toThrowError(
        'schema with key or id "/schemas/customer.create" already exists'
      );
    });
  });

  describe("validateSchema(schema, dataAccessor, dataSetter)", () => {
    test("passthrough data when no schema", async () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      const data = { username: "userA" };
      const dataSetterSpy = jest.fn();
      const dataAccessorSpy = jest.fn().mockReturnValue(data);

      await discoveryApi.validateSchema({
        schema: {},
        dataAccessor: dataAccessorSpy,
        dataSetter: dataSetterSpy,
      });

      expect(dataAccessorSpy).toBeCalledTimes(1);
      expect(dataSetterSpy).toBeCalledTimes(1);
      expect(dataSetterSpy).toBeCalledWith(data);
    });

    test("passthrough data when no schema and async dataAccessor", async () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      const data = { username: "userA" };
      const dataSetterSpy = jest.fn();
      const dataAccessorSpy = jest.fn().mockResolvedValue(data);

      await discoveryApi.validateSchema({
        schema: {},
        dataAccessor: dataAccessorSpy,
        dataSetter: dataSetterSpy,
      });

      expect(dataAccessorSpy).toBeCalledTimes(1);
      expect(dataSetterSpy).toBeCalledTimes(1);
      expect(dataSetterSpy).toBeCalledWith(data);
    });

    test("throws when dataAccessor data are invalid", async () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      const data = { username: "userA" };
      const dataSetterSpy = jest.fn();
      const dataAccessorSpy = jest.fn().mockResolvedValue(data);

      await expect(
        discoveryApi.validateSchema({
          schema,
          dataAccessor: dataAccessorSpy,
          dataSetter: dataSetterSpy,
        })
      ).rejects.toMatchObject({
        errors: expect.arrayContaining([
          createRequiredError("first_name"),
          createRequiredError("last_name"),
          createRequiredError("email"),
          createRequiredError("password"),
        ]),
      });

      expect(dataAccessorSpy).toBeCalledTimes(1);
      expect(dataSetterSpy).not.toBeCalled();
    });

    test("noop when dataAccessor valid without dataSetter", async () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      const data = { username: "userA" };
      const dataAccessorSpy = jest.fn().mockReturnValue(data);

      await discoveryApi.validateSchema({
        schema: {},
        dataAccessor: dataAccessorSpy,
      });

      expect(dataAccessorSpy).toBeCalledTimes(1);
      expect(data).toMatchObject(data);
    });

    test("does not update dataAccessor data in dataSetter", async () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      const data = { username: "userA" };
      const dataAccessorSpy = () => data;
      // @ts-ignore
      const dataSetterSpy = jest.fn().mockImplementation((setterData) => {
        setterData = {};
      });

      await discoveryApi.validateSchema({
        schema: userSchema,
        dataAccessor: dataAccessorSpy,
        dataSetter: dataSetterSpy,
      });

      expect(dataSetterSpy).toBeCalledTimes(1);
      expect(data).not.toBeNull();
    });

    test("passes validatedData to dataSetter", async () => {
      const ajv = DiscoveryApiFactory.createAjv();
      const discoveryApi = new DiscoveryApi(ajv);

      const data = { username: "userA", additional: "a" };
      const dataAccessorSpy = () => data;
      const dataSetterSpy = jest.fn();

      await discoveryApi.validateSchema({
        schema: userSchema,
        dataAccessor: dataAccessorSpy,
        dataSetter: dataSetterSpy,
      });

      expect(dataSetterSpy).toBeCalledTimes(1);
      expect(dataSetterSpy).toBeCalledWith({ username: "userA" });
    });
  });
});
