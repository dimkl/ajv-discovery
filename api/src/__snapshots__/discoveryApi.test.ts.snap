// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscoveryApi getSchemas() returns all non /schemas.* ids as {schemas: AnySchema[]} 1`] = `
{
  "schemas": [],
}
`;

exports[`DiscoveryApi getSchemas() returns all non /schemas.* ids as {schemas: AnySchema[]} 2`] = `
{
  "schemas": [
    {
      "$async": true,
      "$id": "/schemas/customer.create",
      "additionalProperties": false,
      "default": {
        "$async": true,
        "$id": "/schemas/customer.create",
        "additionalProperties": false,
        "description": "Customer create schema",
        "http_method": "POST",
        "http_path": "/customers",
        "properties": {
          "email": {
            "format": "email",
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "first_name": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "last_name": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "password": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
        },
        "required": [
          "first_name",
          "last_name",
          "email",
          "password",
        ],
        "title": "Customer",
        "type": "object",
      },
      "description": "Customer create schema",
      "http_method": "POST",
      "http_path": "/customers",
      "properties": {
        "email": {
          "format": "email",
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "first_name": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "last_name": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "password": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "first_name",
        "last_name",
        "email",
        "password",
      ],
      "title": "Customer",
      "type": "object",
    },
    {
      "$async": true,
      "$id": "/schemas/customer.update",
      "additionalProperties": false,
      "default": {
        "$async": true,
        "$id": "/schemas/customer.create",
        "additionalProperties": false,
        "description": "Customer create schema",
        "http_method": "POST",
        "http_path": "/customers",
        "properties": {
          "email": {
            "format": "email",
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "first_name": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "last_name": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "password": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
        },
        "required": [
          "first_name",
          "last_name",
          "email",
          "password",
        ],
        "title": "Customer",
        "type": "object",
      },
      "description": "Customer create schema",
      "http_method": "POST",
      "http_path": "/customers",
      "properties": {
        "email": {
          "format": "email",
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "first_name": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "last_name": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "password": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "first_name",
        "last_name",
        "email",
        "password",
      ],
      "title": "Customer",
      "type": "object",
    },
  ],
}
`;

exports[`DiscoveryApi getSchemas() returns all registered schemas {schemas: AnySchema[]} 1`] = `
{
  "schemas": [],
}
`;

exports[`DiscoveryApi getSchemas() returns all registered schemas {schemas: AnySchema[]} 2`] = `
{
  "schemas": [
    {
      "$async": true,
      "$id": "/schemas/customer.create",
      "additionalProperties": false,
      "default": {
        "$async": true,
        "$id": "/schemas/customer.create",
        "additionalProperties": false,
        "description": "Customer create schema",
        "http_method": "POST",
        "http_path": "/customers",
        "properties": {
          "email": {
            "format": "email",
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "first_name": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "last_name": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "password": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
        },
        "required": [
          "first_name",
          "last_name",
          "email",
          "password",
        ],
        "title": "Customer",
        "type": "object",
      },
      "description": "Customer create schema",
      "http_method": "POST",
      "http_path": "/customers",
      "properties": {
        "email": {
          "format": "email",
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "first_name": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "last_name": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "password": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "first_name",
        "last_name",
        "email",
        "password",
      ],
      "title": "Customer",
      "type": "object",
    },
    {
      "$async": true,
      "$id": "/schemas/customer.update",
      "additionalProperties": false,
      "default": {
        "$async": true,
        "$id": "/schemas/customer.create",
        "additionalProperties": false,
        "description": "Customer create schema",
        "http_method": "POST",
        "http_path": "/customers",
        "properties": {
          "email": {
            "format": "email",
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "first_name": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "last_name": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
          "password": {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
        },
        "required": [
          "first_name",
          "last_name",
          "email",
          "password",
        ],
        "title": "Customer",
        "type": "object",
      },
      "description": "Customer create schema",
      "http_method": "POST",
      "http_path": "/customers",
      "properties": {
        "email": {
          "format": "email",
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "first_name": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "last_name": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
        "password": {
          "maxLength": 255,
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "first_name",
        "last_name",
        "email",
        "password",
      ],
      "title": "Customer",
      "type": "object",
    },
  ],
}
`;
